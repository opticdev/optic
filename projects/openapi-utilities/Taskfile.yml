# https://taskfile.dev/#/installation
version: "3"

tasks:
  test:
    desc: Test
    cmds:
      - yarn run ws:test
