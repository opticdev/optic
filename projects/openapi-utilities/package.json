{
  "name": "@useoptic/openapi-utilities",
  "license": "MIT",
  "packageManager": "yarn@4.1.0",
  "version": "0.54.3",
  "main": "build/index.js",
  "types": "build/index.d.ts",
  "files": [
    "/build",
    "/src"
  ],
  "scripts": {
    "build": "yarn tsc --build --verbose",
    "test": "ENVIRONMENT=test jest --colors",
    "clean": "rm -rf build"
  },
  "devDependencies": {
    "@babel/core": "^7.17.0",
    "@babel/plugin-transform-runtime": "^7.17.0",
    "@babel/preset-env": "^7.17.0",
    "@babel/preset-typescript": "^7.17.0",
    "@types/babel__core": "^7",
    "@types/babel__preset-env": "^7",
    "@types/is-url": "^1.2.30",
    "@types/js-yaml": "^4.0.5",
    "@types/json-stable-stringify": "^1.0.33",
    "@types/lodash.groupby": "^4.6.6",
    "@types/lodash.isequal": "^4.5.6",
    "@types/lodash.omit": "^4.5.7",
    "@types/node": "^18.0.0",
    "babel-jest": "^29.3.1",
    "jest": "^29.3.1",
    "prettier": "^3.0.0",
    "ts-jest": "^29.0.3",
    "ts-node": "^10.9.1",
    "typescript": "^5.0.0"
  },
  "dependencies": {
    "@useoptic/json-pointer-helpers": "workspace:*",
    "ajv": "^8.6.0",
    "ajv-errors": "~3.0.0",
    "ajv-formats": "~2.1.0",
    "chalk": "^4.1.2",
    "fast-deep-equal": "^3.1.3",
    "is-url": "^1.2.4",
    "js-yaml": "^4.1.0",
    "json-stable-stringify": "^1.0.1",
    "lodash.groupby": "^4.6.0",
    "lodash.isequal": "^4.5.0",
    "lodash.omit": "^4.5.0",
    "node-machine-id": "^1.1.12",
    "openapi-types": "^12.0.2",
    "ts-invariant": "^0.9.3",
    "url-join": "^4.0.1",
    "yaml-ast-parser": "^0.0.43"
  }
}
