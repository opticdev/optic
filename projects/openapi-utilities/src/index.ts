import {
  OpenAPITraverser,
  normalizeOpenApiPath,
} from './openapi3/implementations/openapi3/openapi-traverser';
import { OpenAPIV2, OpenAPIV3, OpenAPIV3_1 } from 'openapi-types';
import { factsToChangelog } from './openapi3/sdk/facts-to-changelog';
export {
  FlatOpenAPIV2,
  FlatOpenAPIV3,
  FlatOpenAPIV3_1,
} from './flat-openapi-types';

import {
  ConceptualLocation,
  IChange,
  IFact,
  ILocation,
  OpenApiFact,
  OpenApiFieldFact,
  OpenApiBodyFact,
  OpenApiBodyExampleFact,
  OpenApiHeaderFact,
  OpenApiKind,
  OpenApiParameterKind,
  OpenApiOperationFact,
  OpenApiRequestParameterFact,
  OpenApiResponseFact,
  OpenApiComponentSchemaExampleFact,
  OperationLocation,
  QueryParameterLocation,
  PathParameterLocation,
  HeaderParameterLocation,
  CookieParameterLocation,
  ResponseHeaderLocation,
  ResponseLocation,
  BodyLocation,
  BodyExampleLocation,
  ComponentSchemaLocation,
  FieldLocation,
  ChangeType,
  OpenApiSpecificationFact,
  OpenApi3SchemaFact,
  FactVariant,
  ChangeVariant,
  OpenApiKindToFact,
  OpenApiRequestFact,
} from './openapi3/sdk/types';
import {
  isFactVariant,
  isChangeVariant,
  isFactOrChangeVariant,
} from './openapi3/sdk/isType';
import {
  sourcemapReader,
  getSourcemapLink,
  GetSourcemapOptions,
} from './openapi3/implementations/openapi3/sourcemap-reader';

export { defaultEmptySpec } from './openapi3/constants';
export * from './openapi3/implementations/openapi3/types';
export * from './utilities/id';

export { SPEC_TAG_REGEXP, sanitizeGitTag } from './specs/tags';

export {
  getFactForJsonPath,
  constructFactTree,
  FactTree,
} from './openapi3/json-path-interpreters';

export { UserError } from './errors';
export { getEndpointsChanges } from './utilities/changelog';
export {
  BodyChange,
  ContentType,
  OpenApiEndpointChange,
  StatusCode,
  groupChangesAndRules,
} from './utilities/group-changes';
export { traverseSpec } from './utilities/traverse-spec';
export { compareChangesByPath } from './utilities/compare-changes-by-path';
export {
  getLabel,
  getOperationsChanged,
  getOperationsChangedLabel,
} from './utilities/count-changed-operations';
export * from './swagger2';
export {
  compareSpecs,
  CompareSpecResults,
} from './compare-specs/compare-specs';
export { groupDiffsByEndpoint } from './openapi3/group-diff';
export { ObjectDiff, typeofDiff, diff, reconcileDiff } from './diff/diff';
export {
  RuleResult,
  Severity,
  textToSev,
  sevToText,
  SeverityTextOptions,
  SeverityText,
} from './results';
export { OpenApiV3Traverser, OAS3 } from './openapi3/traverser';
export { isTruthyStringValue } from './utilities/truthy';
export {
  sourcemapReader,
  GetSourcemapOptions,
  getSourcemapLink,
  OpenApiFact,
  OpenAPITraverser,
  factsToChangelog,
  ConceptualLocation,
  IChange,
  OpenApiFieldFact,
  OpenApiBodyFact,
  OpenApiBodyExampleFact,
  OpenAPIV2,
  OpenAPIV3,
  OpenAPIV3_1,
  OpenApiKind,
  OpenApiParameterKind,
  OpenApiOperationFact,
  OpenApiHeaderFact,
  IFact,
  ILocation,
  OpenApiSpecificationFact,
  OpenApiRequestParameterFact,
  OpenApiResponseFact,
  OpenApiComponentSchemaExampleFact,
  OpenApi3SchemaFact,
  OperationLocation,
  QueryParameterLocation,
  PathParameterLocation,
  HeaderParameterLocation,
  CookieParameterLocation,
  ResponseHeaderLocation,
  ResponseLocation,
  BodyLocation,
  BodyExampleLocation,
  ComponentSchemaLocation,
  FieldLocation,
  ChangeType,
  FactVariant,
  ChangeVariant,
  OpenApiKindToFact,
  isFactVariant,
  isChangeVariant,
  isFactOrChangeVariant,
  OpenApiRequestFact,
};
export { normalizeOpenApiPath };
export * from './coverage/coverage';

export * from './types';

const packageJson = require('../package.json');
export const version = packageJson.version;
