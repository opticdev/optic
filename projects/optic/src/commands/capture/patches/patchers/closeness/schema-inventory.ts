import {
  FlatOpenAPIV3,
  FlatOpenAPIV3_1,
  OAS3,
  OpenAPIV3,
} from '@useoptic/openapi-utilities';
import { jsonPointerHelpers } from '@useoptic/json-pointer-helpers';
import { computeClosenessFromKeyValueTuples, walkSchema } from './closeness';
import { PatchImpact } from '../../patch-operations';
import { PathComponents } from '../../../../oas/operations';
import { Operation } from 'fast-json-patch';
import { SpecPatch, SpecPatches } from '../spec/patches';

type ClosestMatch = { ref: string; percent: number } | undefined;
export class SchemaInventory {
  constructor(private closeness: number = 0.8) {}

  private schemaMap: Map<string, [string, any][]> = new Map();

  addSchemas(
    rootJsonpath: string,
    schemas: { [key: string]: FlatOpenAPIV3.SchemaObject | undefined }
  ) {
    Object.entries(schemas).forEach(([name, schema]) => {
      // this might need to be memoized.
      if (schema)
        this.schemaMap.set(
          jsonPointerHelpers.append(rootJsonpath, name),
          walkSchema(schema)
        );
    });
  }

  findClosest(generatedSchema: FlatOpenAPIV3.SchemaObject): ClosestMatch {
    const thisSchema = walkSchema(generatedSchema);

    let closestMatch: ClosestMatch;
    this.schemaMap.forEach((value, path) => {
      const closeness = computeClosenessFromKeyValueTuples(thisSchema, value);
      // existing match
      if (closestMatch && closeness >= closestMatch.percent) {
        closestMatch = { ref: path, percent: closeness };
      }
      // no match
      if (closeness >= this.closeness) {
        closestMatch = { ref: path, percent: closeness };
      }
    });

    if (closestMatch && closestMatch.percent >= this.closeness) {
      return closestMatch;
    } else {
      return undefined;
    }
  }

  async *refsForAdditions(
    addedPaths: Set<string>,
    spec: FlatOpenAPIV3.Document | FlatOpenAPIV3_1.Document,
    meta: {
      usedExistingRef?: boolean;
    } = {}
  ): SpecPatches {
    if (addedPaths.size === 0) return [];
    const sorted = Array.from(addedPaths).sort();
    let rootSchemasAdded = false;

    for await (let added of sorted) {
      const addedSchema = jsonPointerHelpers.get(spec, added);
      const arrayItems = arrayItemPaths(addedSchema, added);
      const isAPrimitiveSchema =
        addedSchema.type === 'string' ||
        addedSchema.type === 'boolean' ||
        addedSchema.type === 'number' ||
        addedSchema.type === 'integer';

      const isARootSchema =
        jsonPointerHelpers.matches(added, [
          'paths',
          '**',
          '**',
          'responses',
          '**',
          'content',
          '**',
          'schema',
        ]) ||
        // is a root request schema
        jsonPointerHelpers.matches(added, [
          'paths',
          '**',
          '**',
          'requestBody',
          'content',
          '**',
          'schema',
        ]);
      let matchedRoot = false;
      let matchedSub = false;

      if (isARootSchema) {
        const match = this.findClosest(addedSchema);
        // use ref
        if (match) {
          meta.usedExistingRef = true;
          matchedRoot = true;
          const patch: SpecPatch = {
            description: `use $ref ${match.ref}`,
            path: added,
            impact: [PatchImpact.Refactor],
            diff: undefined,
            groupedOperations: [
              {
                op: 'replace',
                path: added,
                value: {
                  $ref: `#${match.ref}`,
                },
              },
            ],
          };
          yield patch;
        }
      }

      if (!matchedRoot) {
        for await (let items of arrayItems) {
          const match = this.findClosest(jsonPointerHelpers.get(spec, items));
          if (match && match.percent > this.closeness) {
            meta.usedExistingRef = true;
            matchedSub = true;
            const patch: SpecPatch = {
              description: `use $ref ${match.ref}`,
              path: items,
              impact: [PatchImpact.Refactor],
              diff: undefined,
              groupedOperations: [
                {
                  op: 'replace',
                  path: items,
                  value: {
                    $ref: `#${match.ref}`,
                  },
                },
              ],
            };
            yield patch;
          }
        }
      }
      //create a root schema
      if (!matchedSub && !matchedRoot && isARootSchema && !isAPrimitiveSchema) {
        const refName = refNameGenerator(added);
        matchedRoot = true;
        let refPath = jsonPointerHelpers.compile([
          'components',
          'schemas',
          refName,
        ]);

        // ensure randomness if we hit a conflict
        let counter = 1;
        while (this.schemaMap.has(refPath)) {
          refPath = jsonPointerHelpers.compile([
            'components',
            'schemas',
            refName + '_' + String(counter),
          ]);
          if (counter > 1000) {
            throw new Error('Could not determine a unique name');
          }
          counter++;
        }

        const schemaOps: Operation[] = [];
        if (!rootSchemasAdded) {
          if (!spec.components) {
            rootSchemasAdded = true;
            schemaOps.push({
              op: 'add',
              path: jsonPointerHelpers.compile(['components']),
              value: {
                schemas: {},
              },
            });
          } else if (!spec.components.schemas) {
            rootSchemasAdded = true;
            schemaOps.push({
              op: 'add',
              path: jsonPointerHelpers.compile(['components', 'schemas']),
              value: {},
            });
          }
        }

        const patch: SpecPatch = {
          description: `create and use $ref for body`,
          path: added,
          impact: [PatchImpact.Refactor],
          diff: undefined,
          groupedOperations: [
            ...schemaOps,
            {
              op: 'add',
              path: refPath,
              value: addedSchema,
            },
            {
              op: 'replace',
              path: added,
              value: {
                $ref: `#${refPath}`,
              },
            },
          ],
        };
        yield patch;
      }
    }
  }
}

function arrayItemPaths(
  initialSchema: FlatOpenAPIV3.SchemaObject,
  initialPath: string
): string[] {
  const results: string[] = [];

  function walk(schema: FlatOpenAPIV3.SchemaObject, path: string, n: number) {
    if (n > 2) return;
    if (OAS3.isArrayType(schema.type) && (schema as any).items) {
      const itemsPath = jsonPointerHelpers.append(path, 'items');
      results.push(itemsPath);
      walk((schema as any).items, itemsPath, n + 1);
    }
    if (OAS3.isObjectType(schema.type) && schema.properties) {
      Object.entries(schema.properties).forEach(([prop, propSchema]) => {
        walk(
          propSchema,
          jsonPointerHelpers.append(path, 'properties', prop),
          n + 1
        );
      });
    }
  }

  walk(initialSchema, initialPath, 0);

  return results;
}

function refNameGenerator(rootBodySchemaPath: string) {
  const [_, path, method, requestBodyOrResponses, ...other] =
    jsonPointerHelpers.decode(rootBodySchemaPath);

  const components = PathComponents.fromPath(path);
  const pathName = components
    .map((i) => capitalizeFirstLetter(i.name.toLowerCase()))
    .join('');

  if (requestBodyOrResponses === 'responses') {
    const statusCode = other[0];

    return `${capitalizeFirstLetter(
      method
    )}${pathName}${statusCode.toString()}ResponseBody`;
  } else if (requestBodyOrResponses === 'requestBody') {
    return `${capitalizeFirstLetter(method)}${pathName}RequestBody`;
  } else {
    return 'SharedComponent';
  }
}

function capitalizeFirstLetter(string) {
  return string.charAt(0).toUpperCase() + string.slice(1).toLowerCase();
}
