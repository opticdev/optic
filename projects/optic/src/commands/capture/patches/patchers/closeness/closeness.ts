import { FlatOpenAPIV3, FlatOpenAPIV3_1 } from '@useoptic/openapi-utilities';
import { jsonPointerHelpers } from '@useoptic/json-pointer-helpers';

export function walkSchema(
  schema: FlatOpenAPIV3_1.SchemaObject,
  trail: string = '',
  tuples: [string, any][] = []
): [string, any][] {
  const {
    // special case these
    oneOf,
    allOf,
    anyOf,
    properties,
    required,
    // skip these
    description,
    // other properties
    ...other
  } = schema;

  if (properties) {
    Object.entries(properties).forEach(([key, value]) => {
      walkSchema(
        value,
        jsonPointerHelpers.append(trail, 'properties', key),
        tuples
      );
    });
  }

  if (allOf) {
    allOf.forEach((schema) => {
      walkSchema(
        schema,
        // intentionally do not add allOf to trail
        trail,
        tuples
      );
    });
  }

  if (anyOf) {
    anyOf.forEach((schema) => {
      walkSchema(
        schema,
        // intentionally do not add allOf to trail
        trail,
        tuples
      );
    });
  }

  if (oneOf) {
    oneOf.forEach((schema) => {
      walkSchema(
        schema,
        // intentionally do not add allOf to trail
        trail,
        tuples
      );
    });
  }

  if (schema.type === 'object') {
    const objectProperties = Object.keys(properties || {}) || [];
    objectProperties.forEach((property) => {
      tuples.push([
        jsonPointerHelpers.append(trail, 'required', property),
        (schema.required || []).includes(property),
      ]);
    });
  }

  Object.entries(other).forEach(([key, value]) => {
    tuples.push([jsonPointerHelpers.append(trail, key), value]);
  });

  return tuples;
}

export function computeClosenessFromKeyValueTuples(
  a: [string, any][],
  b: [string, any][]
): number {
  const aKeys = new Set(a.map((i) => i[0]));
  const bKeys = new Set(b.map((i) => i[0]));
  const keyIntersection = new Set([...aKeys].filter((i) => bKeys.has(i)));
  const keyUnion = new Set([...aKeys, ...bKeys]);

  const intersectSize = keyIntersection.size;
  const keyUnionSize = keyUnion.size;

  if (keyUnionSize === 0) return 0;

  let matchingIntersectSize = 0;

  keyIntersection.forEach((key) => {
    const aEntries = new Set(
      a.filter((i) => i[0] === key).map((i) => JSON.stringify(i[1]))
    );
    const bEntries = new Set(
      b.filter((i) => i[0] === key).map((i) => JSON.stringify(i[1]))
    );
    const intersection = new Set([...aEntries].filter((i) => bEntries.has(i)));
    const union = new Set([...aEntries, ...bEntries]);

    if (intersection.size !== union.size) {
      matchingIntersectSize =
        matchingIntersectSize + intersection.size / union.size;
    } else {
      matchingIntersectSize = matchingIntersectSize + 1;
    }
  });

  return matchingIntersectSize / keyUnionSize;
}
export function computeCloseness(
  oneSchema: FlatOpenAPIV3_1.SchemaObject | FlatOpenAPIV3.SchemaObject,
  otherSchema: FlatOpenAPIV3_1.SchemaObject | FlatOpenAPIV3.SchemaObject
): number {
  return computeClosenessFromKeyValueTuples(
    walkSchema(oneSchema),
    walkSchema(otherSchema)
  );
}
