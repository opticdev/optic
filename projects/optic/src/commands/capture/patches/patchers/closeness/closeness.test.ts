import { describe, it, expect } from '@jest/globals';
import { computeCloseness, walkSchema } from './closeness';
import { FlatOpenAPIV3 } from '@useoptic/openapi-utilities';

const merged: FlatOpenAPIV3.SchemaObject = {
  type: 'object',
  required: ['a', 'b'],
  properties: {
    a: {
      type: 'string',
    },
    b: {
      type: 'number',
    },
  },
};

const oneOf: FlatOpenAPIV3.SchemaObject = {
  oneOf: [{ type: 'string' }, { type: 'number' }],
};
const oneOfSuperset: FlatOpenAPIV3.SchemaObject = {
  oneOf: [{ type: 'string' }, { type: 'number' }, { type: 'object' }],
};

const composed: FlatOpenAPIV3.SchemaObject = {
  allOf: [
    {
      type: 'object',
      required: ['a'],
      properties: {
        a: {
          type: 'string',
        },
      },
    },
    {
      type: 'object',
      required: ['b'],
      properties: {
        b: {
          type: 'number',
        },
      },
    },
  ],
};
const schemaA: FlatOpenAPIV3.SchemaObject = {
  type: 'object',
  required: ['a', 'c'],
  properties: {
    a: {
      type: 'string',
    },

    b: {
      type: 'number',
    },
    c: {
      type: 'array',
      items: {
        type: 'string',
      },
    },
  },
};

const schemaB: FlatOpenAPIV3.SchemaObject = {
  type: 'object',
  required: ['a'],
  properties: {
    a: {
      type: 'string',
    },

    b: {
      type: 'string',
    },
    c: {
      type: 'array',
      items: {
        type: 'string',
      },
    },
    d: {
      type: 'boolean',
    },
  },
};
const schemaC: FlatOpenAPIV3.SchemaObject = {
  type: 'array',
  items: {
    type: 'string',
  },
};

describe('can walk schema', () => {
  it('works for a basic schema', () => {
    expect(walkSchema(schemaA)).toMatchSnapshot();
  });
});

describe('compare closeness', () => {
  it('close score for a close schema', () => {
    expect(computeCloseness(schemaA, schemaB)).toMatchInlineSnapshot(`0.6`);
  });

  it('scores are always commutative for a close schema', () => {
    expect(computeCloseness(schemaA, schemaB)).toEqual(
      computeCloseness(schemaB, schemaA)
    );
  });

  it('far away score for different root types', () => {
    expect(computeCloseness(schemaA, schemaC)).toMatchInlineSnapshot(`0`);
  });

  it('allOf is walked as a merge', () => {
    expect(walkSchema(composed)).toMatchSnapshot();
  });

  it('oneOf shows half if one of its branches is met', () => {
    expect(computeCloseness(oneOf, { type: 'string' })).toMatchInlineSnapshot(
      `0.5`
    );
    expect(computeCloseness(oneOf, { type: 'number' })).toMatchSnapshot(`.5`);
  });

  it('oneOf superset of another shows partial', () => {
    expect(computeCloseness(oneOfSuperset, oneOf)).toMatchInlineSnapshot(
      `0.6666666666666666`
    );
  });

  it('merged and allOfs look the same', () => {
    expect(computeCloseness(merged, composed)).toMatchInlineSnapshot(`1`);
  });
});
