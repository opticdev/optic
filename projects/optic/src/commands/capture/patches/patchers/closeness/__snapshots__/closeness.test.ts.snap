// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can walk schema works for a basic schema 1`] = `
[
  [
    "/properties/a/type",
    "string",
  ],
  [
    "/properties/b/type",
    "number",
  ],
  [
    "/properties/c/type",
    "array",
  ],
  [
    "/properties/c/items",
    {
      "type": "string",
    },
  ],
  [
    "/required/a",
    true,
  ],
  [
    "/required/b",
    false,
  ],
  [
    "/required/c",
    true,
  ],
  [
    "/type",
    "object",
  ],
]
`;

exports[`compare closeness allOf is walked as a merge 1`] = `
[
  [
    "/properties/a/type",
    "string",
  ],
  [
    "/required/a",
    true,
  ],
  [
    "/type",
    "object",
  ],
  [
    "/properties/b/type",
    "number",
  ],
  [
    "/required/b",
    true,
  ],
  [
    "/type",
    "object",
  ],
]
`;

exports[`compare closeness oneOf shows half if one of its branches is met: .5 1`] = `0.5`;
